### PR TITLE
Site Search Page Reset

### PR DESCRIPTION
**Pull request by jtlong3rd**
This PR updates the Site Search UI so that it resets its pagination offset to 0 in two cases:

1. The organism filter's underlying `organismTree` changes (which happens when preferred organisms are toggled).
2. The url-specified `offset` "overshoots" the search results (for example, `offset=12000` in a search result with 11,000 hits).

### CODE REVIEW

File: Client/src/controllers/SiteSearchController.tsx [hunk 1]
```diff
@@ -341,3 +369,32 @@ async function runRequest(requestBody: SiteSearchRequest): Promise<string> {
...
+function useResetOffsetWhenOrgTreeChanges(
+  organismTree: TreeBoxVocabNode | undefined,
+  setOffset: (newOffset: number) => void,
+) {
+  const previousOrgTree = useRef<TreeBoxVocabNode | undefined>(undefined);
+
+  // Maintain a ref to the latest value of "setOffset" so that "setOffset",
+  // which is not referentially stable, can be kept out of the dependency array
+  // of the effect that resets the offset in response to changes in "organismTree"
+  const setOffsetRef = useRef(setOffset);
```
Comment (dmfalke) [inline Client/src/controllers/SiteSearchController.tsx:382]: Are you sure we need to do this? According to the note in the [API docs](https://reactjs.org/docs/hooks-reference.html#usestate) for `useState`:

> React guarantees that setState function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list.

It would be nice if we could remove some of the complexity introduced by the `useRef` hook. Did you observe otherwise?

**Review by dmfalke — APPROVED** [review]
This looks great. I did make an inline comment about the possibility of remove a use of `useRef`. I'll leave it up to you to investigate further, or to just merge as-is.

File: Client/src/controllers/SiteSearchController.tsx (hunk 1 above)
Comment (jtlong3rd) [reply Client/src/controllers/SiteSearchController.tsx:382]: Good question! I agree that avoiding this `setOffsetRef` is ideal. There is a "gotcha" to worry about (which is why I was doing this ref-based approach in the first place).

**TL;DR:** If `setOffset` is added to the dependencies of the "offset-changing effect", an infinite rendering loop occurs when `organismTree` changes.

**Reason:** Our existing `setOffset` callback is misleadingly named - it's actually not a referentially stable callback returned by some `useState`, but rather is derived by a `useCallback`, one of whose dependencies (`updateParams`) changes every time the `location` changes. Since `setOffset` changes the `location`, this can cause infinite loops if `setOffset` is (1) added to an effect's dependency array and (2) invoked in that effect.

One workaround (which I'm about to push) is to add `setOffset` to the "offset-changing effect" dependencies, but to only invoke `setOffset` if `organismTree` was the dependency that changed. I'm also tightening the derivation of `updateParams` so it's "more" referentially stable.

Comment (dmfalke) [reply Client/src/controllers/SiteSearchController.tsx]: My takeaway from what I quoted above is that you don't need to add `setOffset` as a dependency of the "offset-changing effect". However, the fact that it is causing an infinite loop implies that it _is_ changing on re-renders. I'm not sure what I'm missing!

Comment (dmfalke) [reply Client/src/controllers/SiteSearchController.tsx]: Oh, I see where my confusion is coming from! `setOffset` is _not_ a state-setter, as I had assumed, but instead a callback. Sorry for the confusion!

**Review by dmfalke — APPROVED** [review]
